test/e2e/test/generation/generation.go

```diff
@@ -0,0 +1,92 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package generation
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	agentv1alpha1 "github.com/elastic/cloud-on-k8s/pkg/apis/agent/v1alpha1"
+	"github.com/elastic/cloud-on-k8s/test/e2e/test"
+)
+
+func getGeneration(obj client.Object, k *test.K8sClient) (int64, error) {
+	if err := k.Client.Get(context.Background(), types.NamespacedName{Namespace: obj.GetNamespace(), Name: obj.GetName()}, obj); err != nil {
+		return 0, err
+	}
+
+	return obj.GetGeneration(), nil
+}
+
+func getObservedGeneration(obj client.Object, k *test.K8sClient) (int64, error) {
+	if err := k.Client.Get(context.Background(), types.NamespacedName{Namespace: obj.GetNamespace(), Name: obj.GetName()}, obj); err != nil {
+		return 0, err
+	}
+
+	switch v := obj.(type) {
+	case *agentv1alpha1.Agent:
+		return v.Status.ObservedGeneration, nil
+	default:
+		return 0, fmt.Errorf("unsupported type while retrieving status.observedGeneration: %T", v)
+	}
+}
```

Comment:
Assuming we have the following interface `HasObservedGeneration` in `pkg/apis/common/v1/common.go`:

```golang
// +kubebuilder:object:generate=false
type HasObservedGeneration interface {
	client.Object
	GetObservedGeneration() int64
}
```

And that interface implemented by both Elasticsearch and Agent, for example in `pkg/apis/elasticsearch/v1/elasticsearch_types.go`:

```golang
func (es Elasticsearch) GetObservedGeneration() int64 {
	return es.Status.ObservedGeneration
}
```

We could have:

```suggestion
func getObservedGeneration(object commonv1.HasObservedGeneration, k *test.K8sClient) (int64, error) {
	if err := k.Client.Get(context.Background(), k8s.ExtractNamespacedName(object), object); err != nil {
		return 0, err
	}
	return object.GetObservedGeneration(), nil
}
```

Which would allow to have a compile-time check.

Reply:
I added this logic to both the ES, and Kibana PRs we merged before this one, and I updated this to have the same logic.